pkg/ruler/compat.go

```diff
@@ -149,16 +149,29 @@ func MetricsQueryFunc(qf rules.QueryFunc, queries, failedQueries prometheus.Coun
 
 		result, err := qf(ctx, qs, t)
 
-		// We rely on TranslateToPromqlApiError to do its job here... it returns nil, if err is nil.
-		// It returns promql.ErrStorage, if error should be reported back as 500.
-		// Other errors it returns are either for canceled or timed-out queriers (we're not reporting those as failures),
-		// or various user-errors (limits, duplicate samples, etc. ... also not failures).
-		//
-		// All errors will still be counted towards "evaluation failures" metrics and logged by Prometheus Ruler,
-		// but we only want internal errors here.
-		if _, ok := querier.TranslateToPromqlAPIError(err).(promql.ErrStorage); ok {
-			failedQueries.Inc()
+		// We only care about errors returned by underlying Queryable. Errors returned by PromQL engine are "user-errors",
+		// and not interesting here.
```

Comment:
Is this true? For example, if it fails decoding a chunk while iterating seriesset, then it's an underlying error but the error didn't directly come from the `Queryable`.

What's the downside if we remove the check on `QueryableError` at all?

Reply:
Discussed offline. Decoding a chunk is an edge case that is not covered by current implementation. To fix that, we will wrap returned Series and ChunkSeries (from SeriesSet and ChunkSeriesSet), and translate errors returned by their iterators as well.

Reply:
> Discussed offline. Decoding a chunk is an edge case that is not covered by current implementation. To fix that, we will wrap returned Series and ChunkSeries (from SeriesSet and ChunkSeriesSet), and translate errors returned by their iterators as well.

It would be possible to detect errors that occur during iteration of chunks or samples and treat them in the same way as other errors from Queryable, however benchmarks show that performance penalty for wrapping Series/ChunkSeries and their iterators is too big ([~15% more cpu time](https://gist.githubusercontent.com/pstibrany/4ad7e5debec6fe67d51c355aacc52a6e/raw/dcf0620aae0a2f64963d09dfc0848bfe7e86fa52/gistfile2.txt)). Therefore we have decided to not cover this case.

Reply:
☝️  Agree. Thanks for the investigation!